### PR TITLE
Blockchain download and import performance optimizations

### PR DESCRIPTION
**Pull request by arkpar**

### CODE REVIEW

File: libethcore/EthashAux.cpp [hunk 1]
```diff
@@ -87,13 +87,10 @@ ethash_params EthashAux::params(h256 const& _seedHash)
 {
 	Guard l(get()->x_epochs);
 	unsigned epoch = 0;
-	try
+	auto epochIter = get()->m_epochs.find(_seedHash);
+	if (epochIter == get()->m_epochs.end())
 	{
-		epoch = get()->m_epochs.at(_seedHash);
```
Comment (arkpar) [inline libethcore/EthashAux.cpp]: lots of exceptions thrown here

File: libethereum/BlockChain.cpp [hunk 2]
```diff
@@ -609,7 +614,6 @@ ImportRoute BlockChain::import(bytes const& _block, OverlayDB const& _db, Import
 		}
 
 		clog(BlockChainNote) << "   Imported and best" << td << " (#" << bi.number << "). Has" << (details(bi.parentHash).children.size() - 1) << "siblings. Route:" << route;
-		noteCanonChanged();
```
Comment (arkpar) [inline libethereum/BlockChain.cpp]: Please check if this can be removed safely

Comment (gavofyork) [inline libethereum/BlockChain.cpp]: it's critical.

File: libethcore/EthashAux.cpp (hunk 1 above)
Comment (gavofyork) [inline libethcore/EthashAux.cpp]: odd - there should only be at most the number of epochs in the current chain i.e. < 10.

File: libethereum/BlockChain.h [hunk 3]
```diff
@@ -40,6 +42,11 @@
 #include "BlockQueue.h"
 namespace ldb = leveldb;
 
+template <> struct std::hash<std::pair<dev::h256, unsigned>>
+{
+	size_t operator()(const pair<dev::h256, unsigned> &x ) const { return std::hash<dev::h256>()(x.first) ^ std::hash<unsigned>()(x.second); }
```
Comment (gavofyork) [inline libethereum/BlockChain.h]: bad coding style.

**Comment by gavofyork** [comment]
merged.

**Pull request closed without merging**